File: bindings/swift/OpenPano.swift
// OpenPano.swift
// Swift bindings for OpenPano panorama stitching library
//
// Copyright (c) 2024 OpenPano Contributors
// MIT License

import Foundation
#if canImport(UIKit)
import UIKit
#elseif canImport(AppKit)
import AppKit
#endif
import CoreGraphics

// MARK: - Bridging Header Required
// Add to your bridging header:
// #import "openpano_ffi.h"

// MARK: - Stitch Mode

/// Stitching algorithm modes supported by OpenPano.
public enum StitchMode {
    /// Cylindrical projection mode.
    /// Best for standard panoramas where camera rotates around vertical axis.
    /// Requires ordered input and known focal length.
    case cylinder
    
    /// Camera estimation mode with bundle adjustment.
    /// Most flexible - works with arbitrary camera poses and unordered images.
    case estimateCamera
    
    /// Translation mode.
    /// Best for pure translation (document scanning, satellite imagery).
    /// Requires ordered input.
    case translation
    
    /// Convert to C enum value.
    var cValue: OpenpanoMode {
        switch self {
        case .cylinder: return OPENPANO_MODE_CYLINDER
        case .estimateCamera: return OPENPANO_MODE_ESTIMATE_CAMERA
        case .translation: return OPENPANO_MODE_TRANSLATION
        }
    }
}

// MARK: - Log Level

/// Log verbosity levels.
public enum LogLevel: Int32 {
    /// No logging output.
    case none = 0
    /// Only error messages.
    case error = 1
    /// Warnings and errors.
    case warn = 2
    /// Informational messages, warnings, and errors.
    case info = 3
    /// All messages including debug output.
    case debug = 4
    
    /// Convert to C enum value.
    var cValue: OpenpanoLogLevel {
        return OpenpanoLogLevel(rawValue: UInt32(self.rawValue))
    }
}

// MARK: - Errors

/// Errors that can occur during panorama stitching.
public enum OpenPanoError: Error, LocalizedError {
    /// The provided handle is invalid or null.
    case invalidHandle
    /// An invalid argument was provided.
    case invalidArgument(String)
    /// The specified file was not found.
    case fileNotFound(String)
    /// At least 2 images are required for stitching.
    case insufficientImages
    /// The stitching operation failed.
    case stitchingFailed(String)
    /// Memory allocation failed.
    case outOfMemory
    /// Feature detection failed on one or more images.
    case featureDetectionFailed
    /// Feature matching between images failed.
    case matchingFailed
    /// The image format is not supported.
    case unsupportedFormat
    /// The operation was cancelled.
    case cancelled
    /// An unknown error occurred.
    case unknown(Int32)
    
    /// Initialize from C error code.
    init(from error: OpenpanoError) {
        switch error {
        case OPENPANO_ERROR_INVALID_HANDLE: 
            self = .invalidHandle
        case OPENPANO_ERROR_INVALID_ARGUMENT: 
            self = .invalidArgument("")
        case OPENPANO_ERROR_FILE_NOT_FOUND: 
            self = .fileNotFound("")
        case OPENPANO_ERROR_INSUFFICIENT_IMAGES: 
            self = .insufficientImages
        case OPENPANO_ERROR_STITCHING_FAILED: 
            self = .stitchingFailed("")
        case OPENPANO_ERROR_OUT_OF_MEMORY: 
            self = .outOfMemory
        case OPENPANO_ERROR_FEATURE_DETECTION_FAILED: 
            self = .featureDetectionFailed
        case OPENPANO_ERROR_MATCHING_FAILED: 
            self = .matchingFailed
        case OPENPANO_ERROR_UNSUPPORTED_FORMAT:
            self = .unsupportedFormat
        case OPENPANO_ERROR_CANCELLED:
            self = .cancelled
        default: 
            self = .unknown(error.rawValue)
        }
    }
    
    /// Human-readable error description.
    public var errorDescription: String? {
        switch self {
        case .invalidHandle:
            return "Invalid handle"
        case .invalidArgument(let msg):
            return msg.isEmpty ? "Invalid argument" : "Invalid argument: \(msg)"
        case .fileNotFound(let path):
            return path.isEmpty ? "File not found" : "File not found: \(path)"
        case .insufficientImages:
            return "Need at least 2 images for stitching"
        case .stitchingFailed(let msg):
            return msg.isEmpty ? "Stitching failed" : "Stitching failed: \(msg)"
        case .outOfMemory:
            return "Out of memory"
        case .featureDetectionFailed:
            return "Feature detection failed"
        case .matchingFailed:
            return "Feature matching failed"
        case .unsupportedFormat:
            return "Unsupported image format"
        case .cancelled:
            return "Operation cancelled"
        case .unknown(let code):
            return "Unknown error (code: \(code))"
        }
    }
}

// MARK: - Configuration

/// Configuration options for panorama stitching.
///
/// Use the builder pattern to customize settings:
/// ```swift
/// let config = StitcherConfig()
///     .mode(.estimateCamera)
///     .crop(true)
///     .threads(4)
/// ```
public struct StitcherConfig {
    /// Stitching algorithm mode.
    public var mode: StitchMode = .estimateCamera
    
    /// Focal length in 35mm equivalent (primarily for cylinder mode).
    public var focalLength: Float = 36.0
    
    /// Whether input images are in sequential order.
    public var orderedInput: Bool = false
    
    /// Whether to crop result to remove black borders.
    public var cropResult: Bool = true
    
    /// Number of threads (0 = auto-detect).
    public var numThreads: Int = 0
    
    /// Enable lazy image loading to reduce memory.
    public var lazyRead: Bool = false
    
    /// Enable perspective straightening.
    public var straighten: Bool = false
    
    /// Maximum output dimension (0 = no limit).
    public var maxOutputSize: Int = 0
    
    /// Number of bands for multi-band blending (0 = disabled).
    public var multiband: Int = 1
    
    /// Create a new configuration with default values.
    public init() {}
    
    /// Set the stitching mode.
    public func mode(_ mode: StitchMode) -> StitcherConfig {
        var config = self
        config.mode = mode
        return config
    }
    
    /// Set the focal length in 35mm equivalent.
    public func focalLength(_ focal: Float) -> StitcherConfig {
        var config = self
        config.focalLength = focal
        return config
    }
    
    /// Set whether input images are ordered.
    public func ordered(_ ordered: Bool) -> StitcherConfig {
        var config = self
        config.orderedInput = ordered
        return config
    }
    
    /// Set whether to crop the result.
    public func crop(_ crop: Bool) -> StitcherConfig {
        var config = self
        config.cropResult = crop
        return config
    }
    
    /// Set the number of threads.
    public func threads(_ threads: Int) -> StitcherConfig {
        var config = self
        config.numThreads = threads
        return config
    }
    
    /// Enable or disable lazy reading.
    public func lazyRead(_ lazy: Bool) -> StitcherConfig {
        var config = self
        config.lazyRead = lazy
        return config
    }
    
    /// Convert to C config struct.
    func toCConfig() -> OpenpanoConfig {
        var config = OpenpanoConfig()
        openpano_config_default(&config)
        config.mode = mode.cValue
        config.focal_length = focalLength
        config.ordered_input = orderedInput ? 1 : 0
        config.crop_result = cropResult ? 1 : 0
        config.num_threads = Int32(numThreads)
        config.lazy_read = lazyRead ? 1 : 0
        config.straighten = straighten ? 1 : 0
        config.max_output_size = Int32(maxOutputSize)
        config.multiband = Int32(multiband)
        return config
    }
}

// MARK: - Panorama Image

/// A result image from stitching operations.
public class PanoramaImage {
    private var handle: OpenpanoImageHandle?
    
    /// Initialize with a C handle.
    init(handle: OpenpanoImageHandle) {
        self.handle = handle
    }
    
    deinit {
        close()
    }
    
    /// Release resources. Safe to call multiple times.
    public func close() {
        if let handle = handle {
            openpano_image_destroy(handle)
            self.handle = nil
        }
    }
    
    /// Image width in pixels.
    public var width: Int {
        guard let handle = handle else { return 0 }
        var width: Int32 = 0
        openpano_image_dimensions(handle, &width, nil)
        return Int(width)
    }
    
    /// Image height in pixels.
    public var height: Int {
        guard let handle = handle else { return 0 }
        var height: Int32 = 0
        openpano_image_dimensions(handle, nil, &height)
        return Int(height)
    }
    
    /// Image dimensions as (width, height).
    public var dimensions: (width: Int, height: Int) {
        guard let handle = handle else { return (0, 0) }
        var width: Int32 = 0
        var height: Int32 = 0
        openpano_image_dimensions(handle, &width, &height)
        return (Int(width), Int(height))
    }
    
    /// Number of color channels (typically 3 for RGB).
    public var channels: Int {
        guard let handle = handle else { return 0 }
        return Int(openpano_image_channels(handle))
    }
    
    /// Get raw pixel data as RGB bytes.
    /// - Returns: Pixel data in row-major order, 3 bytes per pixel.
    public func pixels() throws -> Data {
        guard let handle = handle else {
            throw OpenPanoError.invalidHandle
        }
        
        var pixels: UnsafePointer<UInt8>?
        var size: Int = 0
        
        let err = openpano_image_data(handle, &pixels, &size)
        guard err == OPENPANO_OK, let pixelData = pixels else {
            throw OpenPanoError(from: err)
        }
        
        return Data(bytes: pixelData, count: size)
    }
    
    #if canImport(UIKit)
    /// Convert to UIImage (iOS/tvOS).
    public func toUIImage() -> UIImage? {
        guard let handle = handle else { return nil }
        
        var pixels: UnsafePointer<UInt8>?
        var size: Int = 0
        
        let err = openpano_image_data(handle, &pixels, &size)
        guard err == OPENPANO_OK, let pixelData = pixels else { return nil }
        
        let dims = dimensions
        guard dims.width > 0 && dims.height > 0 else { return nil }
        
        let data = Data(bytes: pixelData, count: size)
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        
        guard let provider = CGDataProvider(data: data as CFData),
              let cgImage = CGImage(
                width: dims.width,
                height: dims.height,
                bitsPerComponent: 8,
                bitsPerPixel: 24,
                bytesPerRow: dims.width * 3,
                space: colorSpace,
                bitmapInfo: CGBitmapInfo(rawValue: CGImageAlphaInfo.none.rawValue),
                provider: provider,
                decode: nil,
                shouldInterpolate: true,
                intent: .defaultIntent
              ) else { return nil }
        
        return UIImage(cgImage: cgImage)
    }
    #endif
    
    #if canImport(AppKit) && !targetEnvironment(macCatalyst)
    /// Convert to NSImage (macOS).
    public func toNSImage() -> NSImage? {
        guard let handle = handle else { return nil }
        
        var pixels: UnsafePointer<UInt8>?
        var size: Int = 0
        
        let err = openpano_image_data(handle, &pixels, &size)
        guard err == OPENPANO_OK, let pixelData = pixels else { return nil }
        
        let dims = dimensions
        guard dims.width > 0 && dims.height > 0 else { return nil }
        
        let data = Data(bytes: pixelData, count: size)
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        
        guard let provider = CGDataProvider(data: data as CFData),
              let cgImage = CGImage(
                width: dims.width,
                height: dims.height,
                bitsPerComponent: 8,
                bitsPerPixel: 24,
                bytesPerRow: dims.width * 3,
                space: colorSpace,
                bitmapInfo: CGBitmapInfo(rawValue: CGImageAlphaInfo.none.rawValue),
                provider: provider,
                decode: nil,
                shouldInterpolate: true,
                intent: .defaultIntent
              ) else { return nil }
        
        return NSImage(cgImage: cgImage, size: NSSize(width: dims.width, height: dims.height))
    }
    #endif
    
    /// Convert to CGImage.
    public func toCGImage() -> CGImage? {
        guard let handle = handle else { return nil }
        
        var pixels: UnsafePointer<UInt8>?
        var size: Int = 0
        
        let err = openpano_image_data(handle, &pixels, &size)
        guard err == OPENPANO_OK, let pixelData = pixels else { return nil }
        
        let dims = dimensions
        guard dims.width > 0 && dims.height > 0 else { return nil }
        
        let data = Data(bytes: pixelData, count: size)
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        
        guard let provider = CGDataProvider(data: data as CFData) else { return nil }
        
        return CGImage(
            width: dims.width,
            height: dims.height,
            bitsPerComponent: 8,
            bitsPerPixel: 24,
            bytesPerRow: dims.width * 3,
            space: colorSpace,
            bitmapInfo: CGBitmapInfo(rawValue: CGImageAlphaInfo.none.rawValue),
            provider: provider,
            decode: nil,
            shouldInterpolate: true,
            intent: .defaultIntent
        )
    }
    
    /// Save image to file.
    /// - Parameters:
    ///   - path: Output file path
    ///   - quality: JPEG quality (1-100, ignored for PNG)
    public func save(to path: String, quality: Int = 90) throws {
        guard let handle = handle else {
            throw OpenPanoError.invalidHandle
        }
        let err = openpano_image_save(handle, path, Int32(quality))
        guard err == OPENPANO_OK else {
            throw OpenPanoError(from: err)
        }
    }
    
    /// Save image to URL.
    /// - Parameters:
    ///   - url: Output file URL
    ///   - quality: JPEG quality (1-100, ignored for PNG)
    public func save(to url: URL, quality: Int = 90) throws {
        try save(to: url.path, quality: quality)
    }
}

// MARK: - Stitcher

/// Progress callback for stitching operations.
public typealias ProgressCallback = (Float, String) -> Void

/// Panorama stitcher for combining multiple images.
///
/// Example usage:
/// ```swift
/// let stitcher = try Stitcher()
/// try stitcher.addImage(path: "photo1.jpg")
/// try stitcher.addImage(path: "photo2.jpg")
/// let panorama = try stitcher.stitch()
/// try panorama.save(to: "panorama.jpg", quality: 90)
/// ```
public class Stitcher {
    private var handle: OpenpanoStitcherHandle?
    
    /// Create a new stitcher with the specified configuration.
    /// - Parameter config: Configuration options (uses defaults if nil)
    public init(config: StitcherConfig = StitcherConfig()) throws {
        var cConfig = config.toCConfig()
        handle = openpano_stitcher_create(&cConfig)
        guard handle != nil else {
            throw OpenPanoError.outOfMemory
        }
    }
    
    deinit {
        close()
    }
    
    /// Release resources. Safe to call multiple times.
    public func close() {
        if let handle = handle {
            openpano_stitcher_destroy(handle)
            self.handle = nil
        }
    }
    
    /// Add an image file to the stitcher.
    /// - Parameter path: Path to image file
    public func addImage(path: String) throws {
        guard let handle = handle else {
            throw OpenPanoError.invalidHandle
        }
        let err = openpano_add_image_file(handle, path)
        guard err == OPENPANO_OK else {
            if err == OPENPANO_ERROR_FILE_NOT_FOUND {
                throw OpenPanoError.fileNotFound(path)
            }
            throw OpenPanoError(from: err)
        }
    }
    
    /// Add an image from URL.
    /// - Parameter url: URL to image file
    public func addImage(url: URL) throws {
        try addImage(path: url.path)
    }
    
    #if canImport(UIKit)
    /// Add an image from UIImage.
    /// - Parameter image: UIImage to add
    public func addImage(_ image: UIImage) throws {
        guard let handle = handle,
              let cgImage = image.cgImage else {
            throw OpenPanoError.invalidArgument("Invalid UIImage")
        }
        
        try addCGImage(cgImage)
    }
    #endif
    
    #if canImport(AppKit) && !targetEnvironment(macCatalyst)
    /// Add an image from NSImage.
    /// - Parameter image: NSImage to add
    public func addImage(_ image: NSImage) throws {
        guard let handle = handle,
              let cgImage = image.cgImage(forProposedRect: nil, context: nil, hints: nil) else {
            throw OpenPanoError.invalidArgument("Invalid NSImage")
        }
        
        try addCGImage(cgImage)
    }
    #endif
    
    /// Add an image from CGImage.
    /// - Parameter cgImage: CGImage to add
    public func addCGImage(_ cgImage: CGImage) throws {
        guard let handle = handle else {
            throw OpenPanoError.invalidHandle
        }
        
        let width = cgImage.width
        let height = cgImage.height
        let bytesPerRow = width * 3
        
        var pixelData = [UInt8](repeating: 0, count: height * bytesPerRow)
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        
        guard let context = CGContext(
            data: &pixelData,
            width: width,
            height: height,
            bitsPerComponent: 8,
            bytesPerRow: bytesPerRow,
            space: colorSpace,
            bitmapInfo: CGImageAlphaInfo.none.rawValue
        ) else {
            throw OpenPanoError.invalidArgument("Failed to create context")
        }
        
        context.draw(cgImage, in: CGRect(x: 0, y: 0, width: width, height: height))
        
        let err = openpano_add_image_data(
            handle,
            &pixelData,
            Int32(width),
            Int32(height),
            0
        )
        guard err == OPENPANO_OK else {
            throw OpenPanoError(from: err)
        }
    }
    
    /// Add an image from raw RGB pixel data.
    /// - Parameters:
    ///   - pixels: RGB pixel data (3 bytes per pixel)
    ///   - width: Image width
    ///   - height: Image height
    public func addImageData(_ pixels: Data, width: Int, height: Int) throws {
        guard let handle = handle else {
            throw OpenPanoError.invalidHandle
        }
        
        let err = pixels.withUnsafeBytes { (ptr: UnsafeRawBufferPointer) -> OpenpanoError in
            guard let baseAddress = ptr.baseAddress else {
                return OPENPANO_ERROR_INVALID_ARGUMENT
            }
            return openpano_add_image_data(
                handle,
                baseAddress.assumingMemoryBound(to: UInt8.self),
                Int32(width),
                Int32(height),
                0
            )
        }
        
        guard err == OPENPANO_OK else {
            throw OpenPanoError(from: err)
        }
    }
    
    /// Number of images currently added.
    public var imageCount: Int {
        guard let handle = handle else { return 0 }
        return Int(openpano_image_count(handle))
    }
    
    /// Clear all added images.
    public func clear() throws {
        guard let handle = handle else {
            throw OpenPanoError.invalidHandle
        }
        let err = openpano_clear_images(handle)
        guard err == OPENPANO_OK else {
            throw OpenPanoError(from: err)
        }
    }
    
    /// Execute panorama stitching.
    /// - Returns: The stitched panorama image
    public func stitch() throws -> PanoramaImage {
        guard let handle = handle else {
            throw OpenPanoError.invalidHandle
        }
        
        var result: OpenpanoImageHandle?
        let err = openpano_stitch(handle, &result)
        
        guard err == OPENPANO_OK, let imageHandle = result else {
            throw OpenPanoError(from: err)
        }
        
        return PanoramaImage(handle: imageHandle)
    }
    
    /// Cancel an ongoing stitching operation.
    /// Can be called from another thread/queue.
    public func cancel() {
        guard let handle = handle else { return }
        openpano_cancel(handle)
    }
}

// MARK: - Global Functions

/// Get the OpenPano library version string.
public func openPanoVersion() -> String {
    return String(cString: openpano_version())
}

/// Check if JPEG format is supported.
public func hasJPEGSupport() -> Bool {
    return openpano_has_jpeg_support() != 0
}

/// Check if PNG format is supported.
public func hasPNGSupport() -> Bool {
    return openpano_has_png_support() != 0
}

/// Get the number of available CPU cores.
public func cpuCount() -> Int {
    return Int(openpano_get_cpu_count())
}

/// Set the global log level.
public func setLogLevel(_ level: LogLevel) {
    openpano_set_log_level(level.cValue)
}

File: bindings/kotlin/OpenPano.kt
/**
 * OpenPano - Kotlin/JNI Bindings for Panorama Stitching
 *
 * This module provides Kotlin bindings to the OpenPano C++ library for
 * creating panoramic images from multiple overlapping photographs.
 *
 * Example usage:
 * ```kotlin
 * val config = StitcherConfig(
 *     mode = StitchMode.ESTIMATE_CAMERA,
 *     cropResult = true
 * )
 *
 * Stitcher(config).use { stitcher ->
 *     stitcher.addImage("/path/to/photo1.jpg")
 *     stitcher.addImage("/path/to/photo2.jpg")
 *
 *     stitcher.stitch().use { panorama ->
 *         panorama.save("/path/to/output.jpg", quality = 90)
 *         // Or convert to Bitmap
 *         val bitmap = panorama.toBitmap()
 *     }
 * }
 * ```
 *
 * @author OpenPano Contributors
 */
package com.openpano

import android.graphics.Bitmap
import java.io.Closeable
import java.nio.ByteBuffer

/**
 * JNI bridge to OpenPano native library.
 *
 * This object handles loading the native library and declares all JNI methods.
 * Users should not call these methods directly - use the high-level Kotlin API instead.
 */
object OpenPanoJNI {
    init {
        System.loadLibrary("openpano_jni")
    }

    // Stitcher lifecycle
    external fun nativeCreateStitcher(
        mode: Int,
        focalLength: Float,
        orderedInput: Boolean,
        cropResult: Boolean,
        numThreads: Int,
        lazyRead: Boolean,
        straighten: Boolean,
        maxOutputSize: Int,
        multiband: Int
    ): Long

    external fun nativeDestroyStitcher(handle: Long)

    // Image input
    external fun nativeAddImageFile(handle: Long, path: String): Int
    external fun nativeAddImageData(handle: Long, pixels: ByteArray, width: Int, height: Int): Int
    external fun nativeImageCount(handle: Long): Int
    external fun nativeClearImages(handle: Long): Int

    // Stitching
    external fun nativeStitch(handle: Long): Long
    external fun nativeStitchWithProgress(
        handle: Long,
        callback: ProgressCallback?
    ): Long
    external fun nativeCancel(handle: Long): Int

    // Image handling
    external fun nativeDestroyImage(handle: Long)
    external fun nativeImageDimensions(handle: Long): IntArray
    external fun nativeImageChannels(handle: Long): Int
    external fun nativeImageData(handle: Long): ByteArray?
    external fun nativeImageSave(handle: Long, path: String, quality: Int): Int

    // Utility
    external fun nativeVersion(): String
    external fun nativeHasJpegSupport(): Boolean
    external fun nativeHasPngSupport(): Boolean
    external fun nativeGetCpuCount(): Int
    external fun nativeSetLogLevel(level: Int)

    /**
     * Progress callback interface for JNI.
     */
    interface ProgressCallback {
        /**
         * Called during stitching to report progress.
         *
         * @param progress Progress value from 0.0 to 1.0
         * @param stage Description of current operation
         */
        fun onProgress(progress: Float, stage: String)
    }
}

/**
 * Stitching algorithm modes.
 */
enum class StitchMode(val value: Int) {
    /**
     * Cylindrical projection mode.
     * Best for standard panoramas where camera rotates around vertical axis.
     * Requires ordered input and known focal length.
     */
    CYLINDER(0),

    /**
     * Camera estimation mode with bundle adjustment.
     * Most flexible - works with arbitrary camera poses and unordered images.
     */
    ESTIMATE_CAMERA(1),

    /**
     * Translation mode.
     * Best for pure translation (document scanning, satellite imagery).
     * Requires ordered input.
     */
    TRANSLATION(2)
}

/**
 * Log verbosity levels.
 */
enum class LogLevel(val value: Int) {
    /** No logging output */
    NONE(0),
    /** Only error messages */
    ERROR(1),
    /** Warnings and errors */
    WARN(2),
    /** Informational messages, warnings, and errors */
    INFO(3),
    /** All messages including debug output */
    DEBUG(4)
}

/**
 * Configuration options for panorama stitching.
 *
 * @property mode Stitching algorithm mode
 * @property focalLength Focal length in 35mm equivalent (for cylinder mode)
 * @property orderedInput Whether images are in sequential order
 * @property cropResult Whether to crop result to remove black borders
 * @property numThreads Number of threads (0 = auto-detect)
 * @property lazyRead Enable lazy image loading to reduce memory
 * @property straighten Enable perspective straightening
 * @property maxOutputSize Maximum output dimension (0 = no limit)
 * @property multiband Number of bands for multi-band blending (0 = disabled)
 */
data class StitcherConfig(
    val mode: StitchMode = StitchMode.ESTIMATE_CAMERA,
    val focalLength: Float = 36f,
    val orderedInput: Boolean = false,
    val cropResult: Boolean = true,
    val numThreads: Int = 0,
    val lazyRead: Boolean = false,
    val straighten: Boolean = false,
    val maxOutputSize: Int = 0,
    val multiband: Int = 1
)

/**
 * Exceptions thrown by OpenPano operations.
 */
sealed class OpenPanoException(message: String) : Exception(message) {
    /** The provided handle is invalid or null */
    object InvalidHandle : OpenPanoException("Invalid handle")

    /** An invalid argument was provided */
    class InvalidArgument(detail: String = "") : 
        OpenPanoException("Invalid argument${if (detail.isNotEmpty()) ": $detail" else ""}")

    /** The specified file was not found */
    class FileNotFound(path: String = "") : 
        OpenPanoException("File not found${if (path.isNotEmpty()) ": $path" else ""}")

    /** At least 2 images are required for stitching */
    object InsufficientImages : OpenPanoException("Need at least 2 images")

    /** The stitching operation failed */
    class StitchingFailed(detail: String = "") : 
        OpenPanoException("Stitching failed${if (detail.isNotEmpty()) ": $detail" else ""}")

    /** Memory allocation failed */
    object OutOfMemory : OpenPanoException("Out of memory")

    /** Feature detection failed on one or more images */
    object FeatureDetectionFailed : OpenPanoException("Feature detection failed")

    /** Feature matching between images failed */
    object MatchingFailed : OpenPanoException("Feature matching failed")

    /** The image format is not supported */
    object UnsupportedFormat : OpenPanoException("Unsupported format")

    /** The operation was cancelled */
    object Cancelled : OpenPanoException("Operation cancelled")

    /** An unknown error occurred */
    class Unknown(code: Int) : OpenPanoException("Unknown error: $code")

    companion object {
        /**
         * Create an exception from a native error code.
         */
        fun fromCode(code: Int): OpenPanoException = when (code) {
            -1 -> InvalidHandle
            -2 -> InvalidArgument()
            -3 -> FileNotFound()
            -4 -> InsufficientImages
            -5 -> StitchingFailed()
            -6 -> OutOfMemory
            -7 -> FeatureDetectionFailed
            -8 -> MatchingFailed
            -9 -> UnsupportedFormat
            -10 -> Cancelled
            else -> Unknown(code)
        }
    }
}

/**
 * Result panorama image.
 *
 * This class wraps the native image handle and provides methods to access
 * pixel data, convert to Bitmap, or save to file.
 *
 * Always call [close] when done to release native resources.
 */
class PanoramaImage internal constructor(private var handle: Long) : Closeable {

    /**
     * Image dimensions as [width, height].
     */
    val dimensions: Pair<Int, Int>
        get() {
            check(handle != 0L) { "Image already closed" }
            val dims = OpenPanoJNI.nativeImageDimensions(handle)
            return Pair(dims[0], dims[1])
        }

    /**
     * Image width in pixels.
     */
    val width: Int get() = dimensions.first

    /**
     * Image height in pixels.
     */
    val height: Int get() = dimensions.second

    /**
     * Number of color channels (typically 3 for RGB).
     */
    val channels: Int
        get() {
            check(handle != 0L) { "Image already closed" }
            return OpenPanoJNI.nativeImageChannels(handle)
        }

    /**
     * Get raw pixel data as RGB bytes.
     *
     * @return ByteArray containing RGB pixel data in row-major order
     * @throws OpenPanoException if the operation fails
     */
    fun pixels(): ByteArray {
        check(handle != 0L) { "Image already closed" }
        return OpenPanoJNI.nativeImageData(handle)
            ?: throw OpenPanoException.OutOfMemory
    }

    /**
     * Convert to Android Bitmap.
     *
     * @return ARGB_8888 Bitmap, or null if conversion fails
     */
    fun toBitmap(): Bitmap? {
        if (handle == 0L) return null

        val data = OpenPanoJNI.nativeImageData(handle) ?: return null
        val (width, height) = dimensions

        // Convert RGB to ARGB for Bitmap
        val argbData = IntArray(width * height)
        for (i in argbData.indices) {
            val r = data[i * 3].toInt() and 0xFF
            val g = data[i * 3 + 1].toInt() and 0xFF
            val b = data[i * 3 + 2].toInt() and 0xFF
            argbData[i] = (0xFF shl 24) or (r shl 16) or (g shl 8) or b
        }

        return Bitmap.createBitmap(argbData, width, height, Bitmap.Config.ARGB_8888)
    }

    /**
     * Save image to file.
     *
     * Format is determined by file extension:
     * - .jpg, .jpeg: JPEG format
     * - .png: PNG format
     *
     * @param path Output file path
     * @param quality JPEG quality (1-100, ignored for PNG)
     * @throws OpenPanoException if the operation fails
     */
    fun save(path: String, quality: Int = 90) {
        check(handle != 0L) { "Image already closed" }
        val err = OpenPanoJNI.nativeImageSave(handle, path, quality)
        if (err != 0) throw OpenPanoException.fromCode(err)
    }

    /**
     * Release native resources.
     *
     * Safe to call multiple times.
     */
    override fun close() {
        if (handle != 0L) {
            OpenPanoJNI.nativeDestroyImage(handle)
            handle = 0L
        }
    }
}

/**
 * Panorama stitcher for combining multiple images.
 *
 * Example:
 * ```kotlin
 * Stitcher().use { stitcher ->
 *     stitcher.addImage("photo1.jpg")
 *     stitcher.addImage("photo2.jpg")
 *     val panorama = stitcher.stitch()
 *     panorama.save("output.jpg")
 * }
 * ```
 *
 * @param config Configuration options for stitching
 */
class Stitcher(config: StitcherConfig = StitcherConfig()) : Closeable {

    private var handle: Long

    init {
        handle = OpenPanoJNI.nativeCreateStitcher(
            mode = config.mode.value,
            focalLength = config.focalLength,
            orderedInput = config.orderedInput,
            cropResult = config.cropResult,
            numThreads = config.numThreads,
            lazyRead = config.lazyRead,
            straighten = config.straighten,
            maxOutputSize = config.maxOutputSize,
            multiband = config.multiband
        )
        if (handle == 0L) throw OpenPanoException.OutOfMemory
    }

    /**
     * Add an image file to the stitcher.
     *
     * @param path Path to image file
     * @throws OpenPanoException.FileNotFound if file doesn't exist
     */
    fun addImage(path: String) {
        check(handle != 0L) { "Stitcher already closed" }
        val err = OpenPanoJNI.nativeAddImageFile(handle, path)
        if (err != 0) {
            if (err == -3) throw OpenPanoException.FileNotFound(path)
            throw OpenPanoException.fromCode(err)
        }
    }

    /**
     * Add an image from Android Bitmap.
     *
     * @param bitmap Bitmap to add (must be ARGB_8888)
     * @throws OpenPanoException if the operation fails
     */
    fun addImage(bitmap: Bitmap) {
        check(handle != 0L) { "Stitcher already closed" }

        val width = bitmap.width
        val height = bitmap.height
        val pixels = IntArray(width * height)
        bitmap.getPixels(pixels, 0, width, 0, 0, width, height)

        // Convert ARGB to RGB
        val rgbData = ByteArray(width * height * 3)
        for (i in pixels.indices) {
            rgbData[i * 3] = ((pixels[i] shr 16) and 0xFF).toByte()
            rgbData[i * 3 + 1] = ((pixels[i] shr 8) and 0xFF).toByte()
            rgbData[i * 3 + 2] = (pixels[i] and 0xFF).toByte()
        }

        val err = OpenPanoJNI.nativeAddImageData(handle, rgbData, width, height)
        if (err != 0) throw OpenPanoException.fromCode(err)
    }

    /**
     * Add an image from raw RGB pixel data.
     *
     * @param pixels RGB pixel data (3 bytes per pixel)
     * @param width Image width
     * @param height Image height
     * @throws OpenPanoException if the operation fails
     */
    fun addImageData(pixels: ByteArray, width: Int, height: Int) {
        check(handle != 0L) { "Stitcher already closed" }
        val err = OpenPanoJNI.nativeAddImageData(handle, pixels, width, height)
        if (err != 0) throw OpenPanoException.fromCode(err)
    }

    /**
     * Number of images currently added.
     */
    val imageCount: Int
        get() {
            check(handle != 0L) { "Stitcher already closed" }
            return OpenPanoJNI.nativeImageCount(handle)
        }

    /**
     * Clear all added images.
     */
    fun clear() {
        check(handle != 0L) { "Stitcher already closed" }
        val err = OpenPanoJNI.nativeClearImages(handle)
        if (err != 0) throw OpenPanoException.fromCode(err)
    }

    /**
     * Execute panorama stitching.
     *
     * This is a blocking operation that may take significant time.
     *
     * @return The stitched panorama image
     * @throws OpenPanoException if stitching fails
     */
    fun stitch(): PanoramaImage {
        check(handle != 0L) { "Stitcher already closed" }
        val result = OpenPanoJNI.nativeStitch(handle)
        if (result == 0L) throw OpenPanoException.StitchingFailed()
        return PanoramaImage(result)
    }

    /**
     * Execute panorama stitching with progress callback.
     *
     * @param onProgress Called with progress (0-1) and stage description
     * @return The stitched panorama image
     * @throws OpenPanoException if stitching fails
     */
    fun stitch(onProgress: (Float, String) -> Unit): PanoramaImage {
        check(handle != 0L) { "Stitcher already closed" }

        val callback = object : OpenPanoJNI.ProgressCallback {
            override fun onProgress(progress: Float, stage: String) {
                onProgress(progress, stage)
            }
        }

        val result = OpenPanoJNI.nativeStitchWithProgress(handle, callback)
        if (result == 0L) throw OpenPanoException.StitchingFailed()
        return PanoramaImage(result)
    }

    /**
     * Cancel an ongoing stitching operation.
     *
     * Can be called from another thread.
     */
    fun cancel() {
        if (handle != 0L) {
            OpenPanoJNI.nativeCancel(handle)
        }
    }

    /**
     * Release native resources.
     *
     * Safe to call multiple times.
     */
    override fun close() {
        if (handle != 0L) {
            OpenPanoJNI.nativeDestroyStitcher(handle)
            handle = 0L
        }
    }
}

// MARK: - Global Functions

/**
 * Get the OpenPano library version string.
 */
fun openPanoVersion(): String = OpenPanoJNI.nativeVersion()

/**
 * Check if JPEG format is supported.
 */
fun hasJpegSupport(): Boolean = OpenPanoJNI.nativeHasJpegSupport()

/**
 * Check if PNG format is supported.
 */
fun hasPngSupport(): Boolean = OpenPanoJNI.nativeHasPngSupport()

/**
 * Get the number of available CPU cores.
 */
fun cpuCount(): Int = OpenPanoJNI.nativeGetCpuCount()

/**
 * Set the global log level.
 */
fun setLogLevel(level: LogLevel) {
    OpenPanoJNI.nativeSetLogLevel(level.value)
}

